torch/_jit_internal.py

```diff
@@ -421,18 +448,43 @@ def fn(*args, **kwargs):
         else:
             return if_false(*args, **kwargs)
 
-    if if_true.__doc__ is None and if_false.__doc__ is not None:
-        doc = if_false.__doc__
-        if_true.__doc__ = doc
-    elif if_false.__doc__ is None and if_true.__doc__ is not None:
-        doc = if_true.__doc__
-        if_false.__doc__ = doc
-    elif if_false.__doc__ is None and if_true.__doc__ is None:
-        # neither function has a docstring
-        doc = None
+    if_true_signature = inspect.signature(if_true)
+    if_false_signature = inspect.signature(if_false)
+    if if_true_signature.parameters == if_false_signature.parameters:
+        parameters = list(if_true_signature.parameters.values())
     else:
-        raise RuntimeError("only one function can have a docstring")
-    fn.__doc__ = doc
+        if len(if_true_signature.parameters) != len(if_false_signature.parameters):
+            raise RuntimeError(
+                f"The number of parameters in signature of {getattr(if_true, '__name__', repr(if_true))} and "
+                f"{getattr(if_false, '__name__', repr(if_false))} mismatch: "
+                f"{len(if_true_signature.parameters)} != {len(if_false_signature.parameters)}."
+            )
+
+        parameters = [
+            inspect.Parameter(
+                name=extract_attribute(if_true_parameter, if_false_parameter, "name"),
+                kind=extract_attribute(if_true_parameter, if_false_parameter, "kind"),
+                default=extract_attribute(
+                    if_true_parameter, if_false_parameter, "default", sentinel=inspect.Parameter.empty
+                ),
+                annotation=extract_attribute(
+                    if_true_parameter, if_false_parameter, "annotation", sentinel=inspect.Parameter.empty
+                ),
+            )
+            for if_true_parameter, if_false_parameter in zip(
+                if_true_signature.parameters.values(), if_false_signature.parameters.values()
+            )
+        ]
```

Comment:
I'm a bit confused about what this is doing. When would the signatures of the two functions differ? In the case of max_pool2d:
```
def max_pool2d_with_indices(
    input: Tensor, kernel_size: BroadcastingList2[int],
    stride: Optional[BroadcastingList2[int]] = None,
    padding: BroadcastingList2[int] = 0,
    dilation: BroadcastingList2[int] = 1,
    ceil_mode: bool = False,
    return_indices: bool = False
) -> Tuple[Tensor, Tensor]:
    pass
def _max_pool2d(
    input: Tensor, kernel_size: BroadcastingList2[int],
    stride: Optional[BroadcastingList2[int]] = None,
    padding: BroadcastingList2[int] = 0,
    dilation: BroadcastingList2[int] = 1,
    ceil_mode: bool = False,
    return_indices: bool = False
) -> Tensor:
    pass
```
the parameter names, annotations, and types are all the same

Reply:
The signature differs in case one function is annotated with the Python 3 syntax and the other with the Python 2 syntax:

https://github.com/pytorch/pytorch/blob/d6d286f651980a5faaa5631c8f8f8934a49978fd/torch/functional.py#L825-L832

https://github.com/pytorch/pytorch/blob/d6d286f651980a5faaa5631c8f8f8934a49978fd/torch/functional.py#L641-L643

https://github.com/pytorch/pytorch/blob/d6d286f651980a5faaa5631c8f8f8934a49978fd/torch/functional.py#L806-L807

In the second case, the annotations will not be picked up by inspecting the signature. Let me try to fix the annotations and see if we still need this.

Reply:
Should be working out. The new implementation now requires the signatures to match.